src/ImageSharp/Formats/Ico/IcoImageFormatDetector.cs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace SixLabors.ImageSharp.Formats.Ico;
+
+/// <summary>
+/// Detects ico file headers.
+/// </summary>
+public sealed class IcoImageFormatDetector : IImageFormatDetector
+{
+    /// <inheritdoc/>
+    public int HeaderSize { get; } = 4;
+
+    /// <inheritdoc/>
+    public unsafe bool TryDetectFormat(ReadOnlySpan<byte> header, [NotNullWhen(true)] out IImageFormat? format)
+    {
+        fixed (byte* ptr = &header.GetPinnableReference())
```

Comment:
This seems dangerous. You're working too hard to maintain DRY when you should be treating each image format as an entirely separate type.